### ISSUE TITLE
Rpi Zero W / PiCam v2 / v4l2_codec does not work

### ISSUE THREAD

**Issue by henriz** [alfredh/baresip#554]
**Use Case**
Use case: a SIP capable video intercom doorbel, based on a Rpi Zero W, ReSpeaker 2 Mic pi HAT and the pi cam.

**Problem**

I have got this working with video, using v4l2. However, it's terribly slow and doesn't use hardware h264 encoding then, which is a shame, as the Rpi has that. It also has h264 hardware decoding btw.

If I set the config to

video_source         v4l2,/dev/video0
module                  v4l2.so
module                  avcodec.so

this works, abeit very slow. It can't even do 160x120 @ 20fps.

Now if I change it to

video_source         v4l2_codec,/dev/video0
module                  v4l2_codec.so
module                  avcodec.so

It does claim to be working, but the receiving end gets no video. I have tested this with a Grandstream GXV3175 and with linphone.

Note: this goes both for the epoll as poll polling setting. Epoll works fine with v4l2.

**Output using v4l2**

Set video encoder: H264 packetization-mode=0 (500000 bit/s, 25.00 fps)
avcodec: video encoder H264: 25.00 fps, 500000 bit/s, pktsize=1024
video: no video display
v4l2: /dev/video0: found valid V4L2 device (160 x 120) pixfmt=YU12
v4l2: Current input: "Camera 0"
v4l2: current framerate is 30.00 fps
call: stream start (active=1)
audio: start
Set video decoder: H264 packetization-mode=0
avcodec: video decoder H264 ()
777@192.168.178.50: Call established: sip:333@192.168.178.50

**Output using v4l2_codec**

Set video encoder: H264 packetization-mode=0 (500000 bit/s, 25.00 fps)
v4l2_codec: video encoder H264: 25.00 fps, 500000 bit/s, pktsize=1024
video: no video display
v4l2_codec: Driver Caps:
  Driver:        "bm2835 mmal"
  Card:          "mmal service 16.1"
  Bus:           "platform:bcm2835-v4l2"
  Version:       4.0
  Capabilities:  0x85200005
  Formats:
     YU12:    'Planar YUV 4:2:0'
     YUYV:    'YUYV 4:2:2'
     RGB3:    '24-bit RGB 8-8-8'
     JPEG: C  'JFIF JPEG'
  -  H264: C  'H.264'
     MJPG: C  'Motion-JPEG'
     YVYU:    'YVYU 4:2:2'
     VYUY:    'VYUY 4:2:2'
     UYVY:    'UYVY 4:2:2'
     NV12:    'Y/CbCr 4:2:0'
     BGR3:    '24-bit BGR 8-8-8'
     YV12:    'Planar YVU 4:2:0'
     NV21:    'Y/CrCb 4:2:0'
     BGR4:    '32-bit BGRA/X 8-8-8-8'

v4l2_codec: Selected Camera Mode:
  Width:   160
  Height:  120
  PixFmt:  H264
  Field:   1
video: vidcodec 'H264' has no decoder
Set video decoder: H264 packetization-mode=0
777@192.168.178.50: Call established: sip:333@192.168.178.50

_Stats:_

v4l2_codec: encoder stats (keyframes:1, deltaframes:5)

audio           Transmit:     Receive:
packets:           2529         2512
avg. bitrate:      63.4         63.6  (kbit/s)
errors:               0            0
pkt.report:        2492         2480
lost:                 0            9
jitter:             2.6          5.3  (ms)

video           Transmit:     Receive:
packets:             10            0
avg. bitrate:       0.7          0.0  (kbit/s)
errors:               0            0

**Config v4l2**

sip_trans_bsize         128
sip_listen              0.0.0.0:5060
call_local_timeout      120
call_max_calls          4
audio_player            alsa,default
audio_source            alsa,default
audio_alert             alsa,default
audio_level             no
video_source            v4l2,/dev/video0
video_display           fakevideo,nil
video_size              160x120
rtp_tos                 184
rtp_ports               10000-20000
rtcp_enable             yes
rtcp_mux                no
rtp_stats               yes
net_interface           tun0
module_path             /usr/local/lib/baresip/modules
module                  stdio.so
module                  g711.so
module                  alsa.so
module                  v4l2.so
module                 avcodec.so
module_tmp              uuid.so
module_tmp              account.so
module_app              auloop.so
module_app              contact.so
module_app              debug_cmd.so
module_app              menu.so
module_app              vidloop.so
cons_listen             0.0.0.0:5555
http_listen             0.0.0.0:8000
ctrl_tcp_listen         0.0.0.0:4444
evdev_device            /dev/input/event0
ice_turn                no
ice_debug               no

**config v4l2_codec**

sip_trans_bsize         128
sip_listen              0.0.0.0:5060
call_local_timeout      120
call_max_calls          4
audio_player            alsa,default
audio_source            alsa,default
audio_alert             alsa,default
audio_level             no
video_source            v4l2_codec,/dev/video0
video_display           fakevideo,nil
video_size              160x120
rtp_tos                 184
rtp_ports               10000-20000
rtcp_enable             yes
rtcp_mux                no
rtp_stats               yes
net_interface           tun0
module_path             /usr/local/lib/baresip/modules
module                  stdio.so
module                  g711.so
module                  alsa.so
module                  v4l2_codec.so
module                 avcodec.so
module_tmp              uuid.so
module_tmp              account.so
module_app              auloop.so
module_app              contact.so
module_app              debug_cmd.so
module_app              menu.so
module_app              vidloop.so
cons_listen             0.0.0.0:5555
http_listen             0.0.0.0:8000
ctrl_tcp_listen         0.0.0.0:4444
evdev_device            /dev/input/event0
ice_turn                no
ice_debug               no

**System**

Raspberry Pi Zero W, 4.14.70+

**Discussion**

Also, the Rpi supports hardware h264 encoding and decoding, using ffmpeg. So the v4l2_codec should also be capable of decoding. However, in my use case, I only send video, and don't care about the received video. There should also be an option to just dump the received video and not decode it at all, to save CPU. I think this is what you get now when you set v4l2_codec and don't set avcodec, it will complain it has no decoder and litter up the logfile, but it does work. Or better said, it does not work, as v4l2_codec encoding does not seem to work.

I'm a developer myself, with experience in C++/Media/FFMpeg coding. So I can code on it myself, and will start doing so, however, felt obliged to also mention this issue here, maybe Alfred (who knows this code) knows the answer off the bat.

**Relevant info**

-------------------------------------------------------------------------

ffmpeg -f v4l2 -list_formats all -i /dev/video0

ffmpeg version 3.2.10-1~deb9u1+rpt2 Copyright (c) 2000-2018 the FFmpeg developers
built with gcc 6.3.0 (Raspbian 6.3.0-18+rpi1+deb9u1) 20170516
  libavutil      55. 34.101 / 55. 34.101
  libavcodec     57. 64.101 / 57. 64.101
  libavformat    57. 56.101 / 57. 56.101
  libavdevice    57.  1.100 / 57.  1.100
  libavfilter     6. 65.100 /  6. 65.100
  libavresample   3.  1.  0 /  3.  1.  0
  libswscale      4.  2.100 /  4.  2.100
  libswresample   2.  3.100 /  2.  3.100
  libpostproc    54.  1.100 / 54.  1.100
[video4linux2,v4l2 @ 0xc111d0] Raw       :     yuv420p :     Planar YUV 4:2:0 : {32-3280, 2}x{32-2464, 2}
[video4linux2,v4l2 @ 0xc111d0] Raw       :     yuyv422 :           YUYV 4:2:2 : {32-3280, 2}x{32-2464, 2}
[video4linux2,v4l2 @ 0xc111d0] Raw       :       rgb24 :     24-bit RGB 8-8-8 : {32-3280, 2}x{32-2464, 2}
[video4linux2,v4l2 @ 0xc111d0] Compressed:       mjpeg :            JFIF JPEG : {32-3280, 2}x{32-2464, 2}
[video4linux2,v4l2 @ 0xc111d0] Compressed:        h264 :                H.264 : {32-3280, 2}x{32-2464, 2}
[video4linux2,v4l2 @ 0xc111d0] Compressed:       mjpeg :          Motion-JPEG : {32-3280, 2}x{32-2464, 2}
[video4linux2,v4l2 @ 0xc111d0] Raw       : Unsupported :           YVYU 4:2:2 : {32-3280, 2}x{32-2464, 2}
[video4linux2,v4l2 @ 0xc111d0] Raw       : Unsupported :           VYUY 4:2:2 : {32-3280, 2}x{32-2464, 2}
[video4linux2,v4l2 @ 0xc111d0] Raw       :     uyvy422 :           UYVY 4:2:2 : {32-3280, 2}x{32-2464, 2}
[video4linux2,v4l2 @ 0xc111d0] Raw       :        nv12 :         Y/CbCr 4:2:0 : {32-3280, 2}x{32-2464, 2}
[video4linux2,v4l2 @ 0xc111d0] Raw       :       bgr24 :     24-bit BGR 8-8-8 : {32-3280, 2}x{32-2464, 2}
[video4linux2,v4l2 @ 0xc111d0] Raw       :     yuv420p :     Planar YVU 4:2:0 : {32-3280, 2}x{32-2464, 2}
[video4linux2,v4l2 @ 0xc111d0] Raw       : Unsupported :         Y/CrCb 4:2:0 : {32-3280, 2}x{32-2464, 2}
[video4linux2,v4l2 @ 0xc111d0] Raw       :        bgr0 : 32-bit BGRA/X 8-8-8-8 : {32-3280, 2}x{32-2464, 2}

-------------------------------------------------------------------------

v4l2-ctl --list-devices
mmal service 16.1 (platform:bcm2835-v4l2):
        /dev/video0

-------------------------------------------------------------------------

**Comment by henriz** [comment]
I was able to fix this issue without changing code. So one could argue it's not a baresip issue, although I would not completely agree.

The issue is that v4l2 default doesn't send keyframe header data and sends keyframes at low frequency.

To fix this, before starting baresip:

v4l2-ctl --set-ctrl=h264_i_frame_period=10 (This will send a keyframe every 10 frames, you can change this to your liking)

 v4l2-ctl --set-ctrl=repeat_sequence_header=1

Also you need to be using poll. Epoll doesn't work.

When the above commands are executed, you can use v4l2_codec and the pi uses hardware h264 encoding. It makes the difference between barely being able to encode 160x120@10fps, or having no issues encoding 640x480 @ 30fps.

**Comment by alfredh** [comment]
hi @henriz 

really great, thanks for the report and suggested fixes.

I think we should change from fd_listen to using pthreads in the v4l2_codec
module, just like we do in v4l2.so -- this would fix many problems.

regarding the config params (  i_frame_period and repeat_sequence_headers)
I think we should set them programmatically from v4l2_codec module.
are you able to write some code for this ?



Alfred

**Comment by henriz** [comment]
I am, and also willing. I'll do both, however, allow me some time, it's unfamiliar code to me.

Regarding the config params, I feel they should be in the config file. With the above mentioned ctl commands, it will stream to for instance linphone, but not to grandstream phones, due to the video profile. The code sets a profile, however v4l2 uses a different way to set profiles, it has profiles 0 to 4. In order to get the stream working, I also had to change the profile.

So I will also add this profile setting to the config.

Lastly, in my case I'm using one-way video, I don't care about any video the other side sends back. This goes for a lot of ppl who use it for 'doorbell/intercom'-like functionality. I feel this should also be selectable in the config, and I would also want to do that, if you agree.

To keep things consistent, I'd suggest doing this with a fakeencoder device, similar to the fakevideo display device. The main reason for wanting to do this, is CPU/GPU load. There's no reason decoding a stream one can't display (headless units).

**Comment by alfredh** [comment]
hi @henriz 

I made a patch for the pthread part. you can test it if you want.
at least it should solve half the problem.

I am using Logitech C920 camera for testing, it has H264 encoder.

**Comment by henriz** [comment]
Nice. I'll test it, you'll have the results in about an hour.

The raspberry pi GPU can do hardware H264 encoding, so it doesnt really matter if the cam has it or not, as you use /dev/video0, you'll be using the RPI hardware H264 encoding.

I'm using the Raspberry noIR cam v2.1.

**Comment by henriz** [comment]
Tested. After the sysctl commands, I did get image using epoll in the previous version',  but only a keyframe every now and then.

The patched version, streams video fine.

**Comment by henriz** [comment]
Anything else you want me to develop for you? :P ;)

**Comment by alfredh** [comment]
thanks,


now that we have the command documented, it is a good start.
many people ask about this. perhaps we should write a Wiki page.


if the parameters were set from the module, even better. These parameters
include target bitrate, force keyframe, H.264 profile etc.


Alfred

**Comment by alfredh** [comment]
hi @henriz how is it going ? did you have any progress with this ?

**Comment by henriz** [comment]
Hey Alfred,

I tested your patch and it does not fully work. Video is stuttery sometimes. There's something with the keyframes going wrong.

I am going to dive into the code and do the things I talked about, but haven't gotten around to it, was finishing the hardware part. This is finished now, so this weekend I'm starting on the code.

**Comment by alfredh** [comment]
great work, keep up the good work! do you have a picture of the hardware ?

**Comment by henriz** [comment]
The hardware is all off-the-shelf. This also limits the current design. The prototype has 2 rather ugly 'microphone holes' on the front, which I would relocate when I redesign the electronics. Basically I'm going to combine the zero W + Respeaker MIC hat + cam on a single board.

Enclosed pictures of the prototype. It's now fully functional, I'm working on details. Large thing I have still to do, is to compile webRTC for the Pi Zero (=arm6) so I can use the module-echo-cancel of pulseaudio, which will greatly enhance the audio.

Also I noticed baresip doesn't really check the sip connection. If somehow the connection is lost, baresil will not notice untill you actually try to make a call.
![20181029_160608](https://user-images.githubusercontent.com/2608678/47659936-48566900-db96-11e8-985d-83159dd64fb4.jpg)
![20181029_160647](https://user-images.githubusercontent.com/2608678/47659938-48566900-db96-11e8-84d1-a2e652e26317.jpg)
![img-20181028-wa0000](https://user-images.githubusercontent.com/2608678/47659940-48566900-db96-11e8-99b0-cafd9b72407a.jpeg)

**Comment by alfredh** [comment]
very nice photos. the hardware looks quite impressive.

did you get the hardware encoding working?

I added a Wiki page with some details about Video HW acceleration:

https://github.com/alfredh/baresip/wiki/Video-codec-hardware-acceleration

**Comment by alfredh** [comment]
hi @henriz 

any progress on your side ?

if the work has stalled I suggest we close the ticket for now and re-open
it when the work becomes active again.

**Comment by alfredh** [comment]
I will close this ticket now, feel free to re-open it or create a new ticket
if you have time to provide code for this.

**Comment by blademckain** [comment]
@henriz 
can u share your project?
in particular the webrtc package compiled for raspberry pi zero?

**Comment by alexlansovich** [comment]
@henriz Hello. can you help to me with the same task?
